File: 백준/Gold/5972. 택배 배송/택배 배송.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {

    static class Node implements Comparable<Node> {
        int dest, cost;
        Node (int d, int c) {
            this.dest = d;
            this.cost = c;
        }
        @Override
        public int compareTo(Node o) {
            return Integer.compare(this.cost, o.cost);
        }
    }

    static int N, M, INF = 50000 * 1000 + 1;
    static ArrayList<Node>[] edges;
    static int [] dist;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        edges = new ArrayList[N + 1];
        for (int n = 0 ; n < N + 1 ; n++) {
            edges[n] = new ArrayList<Node>();
        }

        for (int m = 0 ; m < M ; m++) {
            st = new StringTokenizer(br.readLine());

            int A = Integer.parseInt(st.nextToken());
            int B = Integer.parseInt(st.nextToken());
            int C = Integer.parseInt(st.nextToken());

            edges[A].add(new Node(B, C));
            edges[B].add(new Node(A, C));
        }

        dist = new int[N + 1];
        Arrays.fill(dist, INF);

        dijkstra();

        System.out.println(dist[N]);
    }

    public static void dijkstra() {
        PriorityQueue<Node> pq = new PriorityQueue<>();
        dist[1] = 0;
        pq.add(new Node(1, 0));

        boolean [] visit = new boolean[N + 1];

        while (!pq.isEmpty()) {
            Node node = pq.poll();

            if (!visit[node.dest]) {
                visit[node.dest] = true;

                for (int i = 0 ; i < edges[node.dest].size() ; i++) {
                    Node next = edges[node.dest].get(i);
                    if (dist[next.dest] > dist[node.dest] + next.cost) {
                        dist[next.dest] = dist[node.dest] + next.cost;
                        pq.add(new Node(next.dest, dist[next.dest]));
                    }
                }
            }
        }
    }
}

File: 백준/Gold/5972. 택배 배송/택배 배송.py
import sys
import heapq
INF = sys.maxsize
N, M = map(int, input().split())
edges = [[] for _ in range(N+1)]

for _ in range(M):
    a, b, c = map(int, input().split())
    edges[a].append((c, b))
    edges[b].append((c, a))


def dijkstra(start):
    queue = []
    heapq.heappush(queue, (0, start))
    dist = [INF] * (N + 1)
    dist[start] = 0
    while queue:
        cost, node = heapq.heappop(queue)
        if dist[node] < cost:
            continue
        for next_cost, next_node in edges[node]:
            if cost + next_cost < dist[next_node]:
                dist[next_node] = cost + next_cost
                heapq.heappush(queue, (dist[next_node], next_node))

    return dist[N]


print(dijkstra(1))